src/ray/raylet/node_manager.cc

```diff
@@ -347,6 +348,62 @@ void NodeManager::Heartbeat() {
   });
 }
 
+void NodeManager::WarnResourceDeadlock() {
+  // Check if any progress is being made on this raylet.
+  for (const auto &task : local_queues_.GetTasks(TaskState::RUNNING)) {
+    // Ignore blocked tasks.
+    if (local_queues_.GetBlockedTaskIds().count(task.GetTaskSpecification().TaskId())) {
+      continue;
+    }
+    // Progress is being made, don't warn.
+    return;
+  }
+
```

Comment:
Maybe add an additional check that there are no resources available on the local node. The `RUNNING` queue can also be empty if there are no workers available (because they haven't started yet or are all died).

Reply:
I thought about that but it could be the actor is not placeable due to its size of resources. We could try checking for that too, but it seems simpler to rely on the periodic delay to avoid the warning firing too often.